lib/src/Pos/Diffusion/Full.hs

```diff
@@ -0,0 +1,445 @@
+
+            sendTx :: TxAux -> d Bool
+            sendTx = give txpLimits $ Diffusion.Txp.sendTx enqueue
+
+            sendUpdateProposal :: UpId -> UpdateProposal -> [UpdateVote] -> d ()
+            sendUpdateProposal = give updateLimits $ Diffusion.Update.sendUpdateProposal enqueue
+
+            sendVote :: UpdateVote -> d ()
+            sendVote = give updateLimits $ Diffusion.Update.sendVote enqueue
+
+            -- FIXME
+            -- SSC stuff has a 'waitUntilSend' motif before it. Must remember to
+            -- investigate that and port it if necessary...
+            -- No, it really should be the logic layer which decides when to send
+            -- things.
+            --
+            -- TODO put these into a Pos.Diffusion.Full.Ssc module.
+            sendSscCert :: MCVssCertificate -> d ()
+            sendSscCert = void . invReqDataFlowTK "ssc" enqueue (MsgMPC OriginSender) (ourStakeholderId logic)
+
+            sendSscOpening :: MCOpening -> d ()
+            sendSscOpening = void . invReqDataFlowTK "ssc" enqueue (MsgMPC OriginSender) (ourStakeholderId logic)
+
+            sendSscShares :: MCShares -> d ()
+            sendSscShares = void . invReqDataFlowTK "ssc" enqueue (MsgMPC OriginSender) (ourStakeholderId logic)
+
+            sendSscCommitment :: MCCommitment -> d ()
+            sendSscCommitment = void . invReqDataFlowTK "ssc" enqueue (MsgMPC OriginSender) (ourStakeholderId logic)
+
+            -- Derive various message size limits from the latest adopted
+            -- block version data (provided by the logic layer).
+
+            sscLimits :: SscLimits d
+            sscLimits = SscLimits
+                -- Copied from existing implementation, formerly in
+                -- Pos.Communication.Limits.
+                { commitmentsNumLimit = succ . ceiling . recip . coinPortionToDouble . bvdMpcThd <$> getAdoptedBVData logic
+                }
+
+            txpLimits :: TxpLimits d
+            txpLimits = TxpLimits
+                { maxTxSize = bvdMaxTxSize <$> getAdoptedBVData logic
+                }
+
+            updateLimits :: UpdateLimits d
+            updateLimits = UpdateLimits
+                -- Copied from existing implementation, formerly in
+                -- Pos.Communication.Limits
+                { updateVoteNumLimit = succ . ceiling . recip . coinPortionToDouble . bvdUpdateVoteThd <$> getAdoptedBVData logic
+                , maxProposalSize = bvdMaxProposalSize <$> getAdoptedBVData logic
+                }
+
+            blockLimits :: BlockLimits d
+            blockLimits = BlockLimits
+                { maxBlockSize = bvdMaxBlockSize <$> getAdoptedBVData logic
+                , maxHeaderSize = bvdMaxHeaderSize <$> getAdoptedBVData logic
+                }
+
+            diffusion :: Diffusion d
+            diffusion = Diffusion {..}
+
+        return DiffusionLayer {..}
+
+  where
+    -- TBD will we need any resources here?
```

Comment:
I don't see any rational usecases of resources acquisition that may be related to diffusion layer. It's one-per-runtime thing and it doesn't interact with anything except some in-memory storage.